### PR TITLE
ET-3578 Postgres module creation

### PR DESCRIPTION
**Pull request by jorge-cr-13**
## What 🕵️ 🔍

- Creating the Postgres module and adding the terragrunt files

## References 📝 🔗

- [Jira task](https://xainag.atlassian.net/jira/software/projects/ET/boards/51?selectedIssue=ET-3578)

## Reviewer checklist

*Reviewer agreement:*

* Reviewers shouldn't review a "Draft Pull Request", but rather wait for it to be marked as "Ready for Review".
* Reviewers do **not** commit or merge the pull request.
* If/when there are changes that need to be applied (applying Terraform or deploying Kubernetes manifests), the **assignee** is the one responsible for doing it.
* Use the labels `not-applied`, `partially-applied` and `fully-applied` so others know the current progress.
  * Remember to change labels accordingly once you start and finish to apply the changes.

### CODE REVIEW

File: aws/b2b_personalization/modules/rds/main.tf [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+resource "aws_db_subnet_group" "private_subnets" {
+  name       = "private-subnets"
+  subnet_ids = toset(var.subnets)
+  tags       = var.tags
+}
+
+resource "aws_db_instance" "postgres" {
+  identifier                      = "${var.name}-postgres-db"
```
Comment (Robert-Steiner) [inline aws/b2b_personalization/modules/rds/main.tf:27]: we can also just use `var.name` and add the `-postgres-db` part in `_env/rds.hcl`
```suggestion
  identifier                      = var.name
```

File: aws/b2b_personalization/modules/rds/main.tf [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+      rule        = "postgresql-tcp"
+      description = "Postgres port"
+    }
+  ]
+
+  tags = var.tags
+}
+
+resource "aws_db_subnet_group" "private_subnets" {
+  name       = "private-subnets"
```
Comment (Robert-Steiner) [inline aws/b2b_personalization/modules/rds/main.tf:21]:
```suggestion
  name       = "${var.name}-subnet-group"
```

File: aws/b2b_personalization/modules/rds/main.tf [hunk 3]
```diff
@@ -0,0 +1,50 @@
+module "security_group" {
+  source = "git::https://github.com/terraform-aws-modules/terraform-aws-security-group?ref=v4.16.0"
+
+  name        = "db-security-group-${var.name}"
```
Comment (Robert-Steiner) [inline aws/b2b_personalization/modules/rds/main.tf:4]:
```suggestion
  name        = "${var.name}-sg"
```

File: aws/b2b_personalization/modules/rds/main.tf [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  db_subnet_group_name            = aws_db_subnet_group.private_subnets.name
+  instance_class                  = var.instance
+  allocated_storage               = var.allocated_storage
+  max_allocated_storage           = var.max_allocated_storage
+  engine                          = "postgres"
+  engine_version                  = var.postgres_version
+  skip_final_snapshot             = true
+  publicly_accessible             = false
+  vpc_security_group_ids          = [module.security_group.security_group_id]
+  multi_az                        = true
```
Comment (Robert-Steiner) [inline aws/b2b_personalization/modules/rds/main.tf:39]: I would like to have the `skip_final_snapshot` and `multi_az` configurable. for dev we can turn off `multi_az` and for prod we should turn on `skip_final_snapshot`. for prod we also need to be able to configure [backup_retention_period](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/db_instance#backup_retention_period). currently it is set to `0` which means no backups

File: aws/global/kms/backend.tf [hunk 5]
```diff
@@ -0,0 +1,24 @@
+terraform {
```
Comment (Robert-Steiner) [inline aws/global/kms/backend.tf:1]: i would like to have this as a module that we deploy in each env later

File: aws/b2b_personalization/modules/rds/main.tf [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+  enabled_cloudwatch_logs_exports = ["postgresql", "upgrade"]
+  tags                            = var.tags
+}
+
+resource "aws_ssm_parameter" "postgres_url" {
+  name        = "${var.name}-postgres-url"
```
Comment (Robert-Steiner) [inline aws/b2b_personalization/modules/rds/main.tf:45]:
```suggestion
  name        = "postgres/${var.name}/url"
```

File: aws/b2b_personalization/db_secrets.yaml [hunk 7]
```diff
@@ -0,0 +1,18 @@
+dev:
```
Comment (Robert-Steiner) [inline aws/b2b_personalization/db_secrets.yaml:1]: wow sops setup worked without any issues on my system :)  
we keep the file a bit more general. for example we can name it `secrets.yaml` and move it into the `dev` folder.
```suggestion
postgres_db:
```

File: aws/b2b_personalization/modules/rds/main.tf [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+      description = "Postgres port"
+    }
+  ]
+
+  tags = var.tags
+}
+
+resource "aws_db_subnet_group" "private_subnets" {
+  name       = "private-subnets"
+  subnet_ids = toset(var.subnets)
```
Comment (Robert-Steiner) [inline aws/b2b_personalization/modules/rds/main.tf:22]: do we need to use `toset` here?

File: aws/b2b_personalization/modules/rds/main.tf [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+  identifier                      = "${var.name}-postgres-db"
+  username                        = var.postgres_admin_username
+  password                        = var.postgres_admin_password
+  db_subnet_group_name            = aws_db_subnet_group.private_subnets.name
+  instance_class                  = var.instance
+  allocated_storage               = var.allocated_storage
+  max_allocated_storage           = var.max_allocated_storage
+  engine                          = "postgres"
+  engine_version                  = var.postgres_version
+  skip_final_snapshot             = true
```
Comment (Robert-Steiner) [inline aws/b2b_personalization/modules/rds/main.tf:36]:
```suggestion
  skip_final_snapshot             = true
  storage_encrypted = true
```

File: aws/global/kms/backend.tf (hunk 5 above)
Comment (jorge-cr-13) [reply aws/global/kms/backend.tf:1]: Ok, I'll create this in another PR. :)

File: aws/b2b_personalization/_env/rds.hcl [hunk 10]
```diff
@@ -0,0 +1,36 @@
+locals {
+  secrets  = yamldecode(sops_decrypt_file(find_in_parent_folders("db_secrets.yaml")))
```
Comment (Robert-Steiner) [inline aws/b2b_personalization/_env/rds.hcl:2]:
```suggestion
  secrets  = yamldecode(sops_decrypt_file(find_in_parent_folders("secrets.yaml")))
```

File: aws/b2b_personalization/_env/rds.hcl [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+  source = "../../modules//rds"
+}
+
+inputs = {
+  name    = "${local.env_vars.locals.name}-postgres-db"
+  vpc_id  = dependency.vpc.outputs.vpc_id
+  subnets = dependency.vpc.outputs.subnet_private_ids
+
+  postgres_admin_username = local.secrets.dev.username
+  postgres_admin_password = local.secrets.dev.password
```
Comment (Robert-Steiner) [inline aws/b2b_personalization/_env/rds.hcl:26]:
```suggestion
  postgres_admin_username = local.secrets.postgres_db.username
  postgres_admin_password = local.secrets.postgres_db.password
```

File: aws/global/kms/backend.tf (hunk 5 above)
Comment (Robert-Steiner) [reply aws/global/kms/backend.tf:1]: might be that this will recreate the key but it doesn't matter at the moment since nothing is deployed or in use

**Review by Robert-Steiner — COMMENTED** [review]
just two small things rest looks good :)

**Review by joecks — APPROVED** [review]

**Pull request merged**